Add cpp-test plugin

big-guy opened the pull request:
### Context
<!--- Why do you believe many users will benefit from this change? -->
<!--- Link to relevant issues or forum discussions here -->

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [ ] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppTestConventionPlugin.java:
```diff
@@ -0,0 +1,121 @@
+        final TaskContainer tasks = project.getTasks();
+
+        // TODO: What should this name be?
+        CppTestSuite component = objectFactory.newInstance(DefaultCppTestSuite.class, "test", objectFactory, fileOperations, configurations);
```
should probably also be made visible as an extension, like the other plugins.

big-guy replied:
done

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppTestConventionPlugin.java:
```diff
@@ -0,0 +1,121 @@
+            }
+        });
+
+        tasks.getByName("check", new Action<Task>() {
```
not just `tasks.getByName("check").dependsOn(testTask)`?

big-guy replied:
done

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppTestConventionPlugin.java:
```diff
@@ -0,0 +1,121 @@
+ * @since 4.4
+ */
+@Incubating
+public class CppTestConventionPlugin implements Plugin<ProjectInternal> {
```
needs a unit test

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/internal/DefaultCppTestSuite.java:
```diff
@@ -0,0 +1,32 @@
+
+import javax.inject.Inject;
+
+public class DefaultCppTestSuite extends DefaultCppApplication implements CppTestSuite {
```
should be `extends DefaultCppComponent` rather than `extends DefaultCppApplication`.

big-guy replied:
done

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/CppTestSuite.java:
```diff
@@ -0,0 +1,37 @@
+ */
+@Incubating
+public interface CppTestSuite extends CppComponent {
+    Property<CppComponent> getTestedComponent();
```
needs some javadoc

big-guy replied:
I can remove this from the public API too for now.

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/CppTestSuite.java:
```diff
@@ -0,0 +1,37 @@
+     * {@inheritDoc}
+     */
+    @Override
+    CppTestExecutable getDevelopmentBinary();
```
what does this mean?

big-guy replied:
This will be the test executable, not the component-under-test's binary.  There's only a single binary for unit tests right now.

adammurdoch commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/googletest/plugins/GoogleUnitTestPlugin.java:
```diff
@@ -0,0 +1,55 @@
+import org.gradle.nativeplatform.test.cpp.plugins.CppUnitTestBasePlugin;
+import org.gradle.nativeplatform.test.tasks.RunTestExecutable;
+
+public class GoogleUnitTestPlugin implements Plugin<Project> {
```
This doesn't really have anything to do with google test. We just want a generic 'run this executable' plugin. Then, later, we would add a sibling to this plugin with better integration for google test.

big-guy replied:
Yep, there was another commit after this that Gary and I worked on after you left that created a `GoogleTest` task and scraped the output from google test execution.

I pushed that experiment to a branch and detangle it from here, so we just have a `cpp-unit-test` plugin that adds a `RunTestExecutable` and wires everything together.  We can use this to make a native-sample to use googleTest or catch or whatever with this.

adammurdoch commented on buildSrc/src/main/groovy/org/gradle/build/docs/ExtractSnippetsTask.groovy:
```diff
@@ -15,38 +15,68 @@
-        return super.getSource()
+        def source = project.fileTree(samples)
+        source.with {
+            exclude 'userguideOutput/**'
```
seems odd to bake this stuff into a task implementation.

big-guy replied:
I moved this to another PR.  The task wasn't cacheable and caused us problems when we wanted to build a new variant of googletest.

lacasseio commented on subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppUnitTestBasePlugin.java:
```diff
@@ -0,0 +1,90 @@
+
+        project.getPlugins().withType(CppLibraryPlugin.class, projectConfiguration);
+        // TODO: We will get symbol conflicts with executables since they already have a main()
+        project.getPlugins().withType(CppExecutablePlugin.class, projectConfiguration);
```
Could we simply trigger on `CppBasePlugin` instead?

big-guy replied:
No, because we always add the base plugin when applying this plugin.  There's no "main" component defined until we have a CppApplication or CppLibrary created by one of the other plugins

big-guy commented on subprojects/language-native/src/main/java/org/gradle/language/cpp/CppBinary.java:
```diff
@@ -63,4 +63,10 @@
+     * TODO: Files before linking
+     * @return
+     */
+    // FileCollection getIntermediateFiles();
```
yep, this was leftovers.

big-guy commented on subprojects/language-native/src/main/java/org/gradle/language/cpp/CppComponent.java:
```diff
@@ -89,4 +89,6 @@
      */
     CppBinary getDevelopmentBinary();
+
+    FileCollection getAllHeaderDirs();
```
I think we can get by with not exposing this for now.

big-guy commented:
I've simplified this a bit and removed the google-test specific things that were wrapped up in it.  Could you take another look @adammurdoch?

I've kicked off another CI build for it.
